internal/hcsoci/resources.go

```diff
@@ -34,151 +31,60 @@ type Resources struct {
 	// For WCOW, this will be under wcowRootInUVM. For LCOW, this will be under
 	// lcowRootInUVM, this will also be the "OCI Bundle Path".
 	containerRootInUVM string
-
-	// layers is an array of the layer folder paths which have been mounted either on
-	// the host in the case or a WCOW Argon, or in a utility VM for WCOW Xenon and LCOW.
-	layers []string
-
-	// vsmbMounts is an array of the host-paths mounted into a utility VM to support
-	// (bind-)mounts into a WCOW v2 Xenon.
-	vsmbMounts []string
-
-	// pipeMounts is an array of the host-paths of named pipes mounted into a utility
-	// VM.
-	pipeMounts []string
-
-	// plan9Mounts is an array of all the host paths which have been added to
-	// an LCOW utility VM
-	plan9Mounts []*uvm.Plan9Share
-
-	// netNS is the network namespace
-	netNS string
-
-	// networkEndpoints is the list of network endpoints used by the container
-	networkEndpoints []string
-
+	netNS              string
 	// createNetNS indicates if the network namespace has been created
 	createdNetNS bool
-
 	// addedNetNSToVM indicates if the network namespace has been added to the containers utility VM
 	addedNetNSToVM bool
-
-	// scsiMounts is an array of the vhd's mounted into a utility VM to support
-	// scsi device passthrough.
-	scsiMounts []scsiMount
-
-	// vpciDevices is an array of vpci device names
-	vpciDevices []string
+	// resources is an array of the resources associated with a container
+	resources []ResourceCloser
 }
 
-type scsiMount struct {
-	// path is the host path to the vhd that is mounted.
-	path string
-	// autoManage if `true` means that on cleanup, the runtime should
-	// automatically delete this vhd.
-	autoManage bool
+// ResourceCloser is a generic interface for the releasing of a resource. If a resource implements
+// this interface(which they all should), freeing of that resource should entail one call to
+// <resourceName>.Release(ctx, all)
+type ResourceCloser interface {
+	Release(context.Context, bool) error
```

Comment:
I think we should try to refactor the code such that we don't pass the `all` bool here. I think it's a leaky abstraction to require each resource to know under which conditions it should actually be cleaned up. Rather we should group the `ResourceCloser`s in `Resources` into categories so that it's actually `ReleaseResources` which makes the decision of which resources to clean up, and if `Release` is called on a resource then it is for sure released.

Reply:
We can just extend on the switch statement right now and in the default case check if all is False and then if the type is anything but *ImageLayers don't call release. I'd like to get @jstarks take on this though as we discussed this implementation.

Reply:
Actually we can't do what I said above and check in default if it's *ImageLayers without using reflection as the type has to be checked as the case. We can just check for either network or layers and if the default case gets hit these should be all the types the all bool is concerned with. Kind of redundant looking and there's probably a better way but this gets the point across. 
```go
for _, resource := range r.resources {
		if resource != nil {
			switch resource.(type) {
			case *uvm.NetworkEndpoints:
				if r.createdNetNS {
					if err := resource.Release(ctx); err != nil {
						return err
					}
					r.createdNetNS = false
				}
			case *ImageLayers:
				if err := resource.Release(ctx); err != nil {
					return err
				}
			default:
				if all {
					if err := resource.Release(ctx); err != nil {
						return err
					}
				}
			}
		}
	}
```